transformations/antonyms_substitute/README.md

```diff
@@ -0,0 +1,47 @@
+# Antonyms Substitute
+This transformation aims to substitute words with their antonyms, which facilitates the diversity of the content.
+
+Author name:
+- Zhexiong Liu (zhexiong@cs.pitt.edu, University of Pittsburgh)
+- Jing Zhang (jing.zhang2@emory.edu, Emory University)
+
+## What type of a transformation is this?
+This transformation could introduce semantic diversity by adding antonyms. Specifically, it will help revert the semantics to an opposite position or uses double negation to express the similar semantics.
```

Comment:
Could you give a few examples as to when the the semantics is reverted and when similar semantics result from the transformation? Just to make the readme better. Also, for its use in sentiment analysis, the transformation has to guarantee that the label is either preserved or altered. Currently, the transformation does not provide this. e.g. even if only one substitution is made, is it guaranteed that the label will be the same? e.g. That restaurant is fantastically awful. --> That restaurant is awfully fantastic. is a label change. But not all transformations will lead to label change, as with the examples in the test cases as well. So, without the label effects it is hard to do automatic augmentation for text classification. @kaustubhdhole If the targeted task is text classification, is it necessary to make the transformation label preserving?

Reply:
Hi @vyraun thanks for your review! I am not sure if we need to preserve the label while doing the transformation, but we could generally assume that an even number of transforms will remain the semantics in terms of sentiment; however, an odd number of transforms will revert the semantics. In your examples, yes it should have label changes but the case was not representative in most cases as the awful and fantastic are anonymous themself, which was not usual in most cases. We could develop a fine-grained algorithm focusing on ADJ or ADV only to ensure the preserve of the label but I would assume it was beyond the goal of this project.

Reply:
So, if the meaning changes, it is important to have a way to track that. Either by modelling your transformation with a label eg. positive vs negative sentiment, etc. or else it can be hard to use the transformation in a subsequent experiment. Using a sentenceoperation would mean you could basically substitute one sentence for another in ideally any setting. Either you can mention specific tasks (in the README as well as add them in the TaskTypes class) which would be agnostic to antonym changes. Probably, it is hard to think of that since "antonym" theoretically means "opposite meaning" and depending on where in the sentence the antonym is reversed, it can be hard to track the consistency with a task. Another way to be slightly safer (as compared to the above approach) but not ideal again is to use your approach in a question answering setting where you can probably make changes to only words common to both the context and the question?

Reply:
Hi @kaustubhdhole, thanks for your great suggestion! We update our code to ensure the transformation only applies to the sentence that has an even number of revertable words. We call it double negation, which means it would revert back the sentence semantics to its original (e.g. more interested -> less uninterested; often reputable -> infrequently disreputable;). We evaluated this transformation and achieved 83% accuracy on the sentiment analysis task. All code and readme and test cases were updated. I think all our previous concerns are dismissed.

Reply:
Thanks @ZhexiongLiu , guaranteeing label preservation will make the transformation more usable. However, as in the counter-example I gave above, double negation doesn't guarantee label preservation: "That restaurant is fantastically awful." --> "That restaurant is awfully fantastic.", and of course, there could be more contexts when the general rule is not followed, e.g. "Ram is a talented and skilled archer" --> "Ram is a untalented and unskilled archer". Could you add a condition that if the two words in the applied double negation are already antonyms or synonyms, the transformation isn't applied?

Reply:
Thanks @vyraun , I've added the condition that does not apply double negation if two words are already antonyms or synonyms. Now the IMDB results boosted to 93. Great suggestions!